File: bindings/examples/shopping.cpp
/* The C++ guide's example: compile a grammar from a .raku file, parse text,
 * and move the results into C++. Unlike the other hosts, C++ LINKS against
 * librakupp instead of dlopen'ing it. From a checkout, on macOS:
 *
 *   c++ -std=c++17 -Isrc bindings/examples/shopping.cpp \
 *       build/librakupp.dylib -Wl,-rpath,$PWD/build -o shopping && ./shopping
 *
 * On Linux:  c++ -std=c++17 -Isrc bindings/examples/shopping.cpp \
 *       -Lbuild -lrakupp -Wl,-rpath,$PWD/build -lpthread -o shopping
 *
 * Against an INSTALLED rakupp the include is <rakupp/grammar.hpp> and the
 * flags are just -lrakupp.
 */
#include "grammar.hpp"

#include <iostream>

int main() {
    auto g = rakupp::Grammar::from_file("bindings/examples/shopping.raku",
                                        "Shopping", "ShoppingActions");

    // parse returns std::nullopt if the grammar does not match
    auto m = g.parse("milk=2\nbread=1\neggs=12\n");
    if (!m) {
        std::cerr << "no match\n";
        return 1;
    }

    auto items = (*m)["item"];
    std::cout << items.size() << " items\n";
    for (size_t i = 0; i < items.size(); i++)   // lazy: one engine call per leaf
        std::cout << items[i]["name"].str() << " x " << items[i]["qty"].int_() << "\n";

    // ShoppingActions made this
    std::cout << "total, computed in Raku: " << m->made().int_() << "\n";

    rakupp::Tree all = m->tree();               // the whole match, eagerly
    std::cout << "items in the tree: " << all.map().at("item").list().size() << "\n";

    // parse gives std::nullopt on a non-match; parse_or_throw diagnoses it
    try {
        g.parse_or_throw("milk=2\nbread=lots\n");
    }
    catch (const rakupp::ParseError& e) {
        std::cout << e.what() << "\n";
    }
    return 0;
}   // m's destructor unroots the engine value

File: bindings/rust/examples/shopping.rs
// The Rust guide's example: compile a grammar from a .raku file, parse text,
// and move the results into Rust. From a checkout (BUILD = a directory
// holding librakupp):
//
//   RAKUPP_LIB_DIR=$BUILD cargo run --example shopping
//
// (run from bindings/rust, or add --manifest-path bindings/rust/Cargo.toml)

use rakulang::{Error, Grammar, Tree};

fn main() -> Result<(), Error> {
    let grammar = concat!(env!("CARGO_MANIFEST_DIR"), "/../examples/shopping.raku");
    let g = Grammar::from_file(grammar, "Shopping", "ShoppingActions")?;

    // parse returns Ok(None) if the grammar does not match
    let m = g.parse("milk=2\nbread=1\neggs=12\n", "")?.expect("the list matches");

    let items = m.get("item");
    println!("{} items", items.len()?);
    for i in 0..items.len()? {
        // lazy: one engine call per leaf
        let item = items.at(i);
        println!("{} x {}", item.get("name").str()?, item.get("qty").int()?);
    }

    // ShoppingActions made this
    if let Tree::Int(total) = m.made()? {
        println!("total, computed in Raku: {}", total);
    }

    println!("as plain Rust data: {:?}", m.tree()?);

    // parse gives Ok(None) on a non-match; parse_strict diagnoses it instead
    if let Err(Error::Parse { line, col, rule, .. }) = g.parse_strict("milk=2\nbread=lots\n", "") {
        println!("line {} column {} while trying <{}>", line, col, rule);
    }

    Ok(())
} // m drops here — the rooted engine value unroots itself
